### PR TITLE
Lag-interface replaced and overridden states

### PR DESCRIPTION
**Pull request by mingjunzhang2019**
SUMMARY
This is to add/remedy 'replaced' and 'overridden' states for sonic_lag_interfaces module.

GitHub Issues
N/A

ISSUE TYPE
- Feature Pull Request

COMPONENT NAME
sonic_lag_interfaces

OUTPUT
- regression test result:
[regression-2023-03-29-10-14-49.html.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/11104176/regression-2023-03-29-10-14-49.html.pdf)

ADDITIONAL INFORMATION
- regression test log: 
[regression_test.log](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/11104181/regression_test.log)

Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [x] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

How Has This Been Tested?
Run the tests on several real SONIC switch systems.

### CODE REVIEW

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 1]
```diff
@@ -208,20 +208,30 @@ def _state_overridden(self, want, have, diff_members, diff_portchannels):
...
-        commands_overridden, requests_overridden = self.template_for_lag_deletion(have, delete_members, delete_portchannels, "overridden")
-        requests.extend(requests_overridden)
-        commands.extend(commands_overridden)
+
+        deleted_po_list = list()
+        for i in delete_list:
+            list_obj = search_obj_in_list(i['name'], want, "name")
+            if not list_obj:
+                deleted_po_list.append(i)
+
```
Comment (ArunSaravananBalachandran) [inline plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:226]: Determining the contents of the replaced_list and deleted_po_list could be simplied as below:
```python
        replaced_list = []
        deleted_po_list = []
        for i in delete_list:
            list_obj = search_obj_in_list(i['name'], want, 'name')
            if not list_obj:
                deleted_po_list.append(i)
            elif 'members' in i.keys():
                replaced_list.append(i)
```

File: tests/regression/roles/sonic_lag_interfaces/defaults/main.yml [hunk 2]
```diff
@@ -77,13 +79,49 @@ tests:
...
+        - name: portchannel 40
+          members:
+            interfaces:
+              - member: "{{ interface5 }}"
+        - name: po41
+          members:
+            interfaces:
+              - member: "{{ interface6 }}"
+  - name: test_case_07
+    description: Overridden portchannel configuration
```
Comment (ArunSaravananBalachandran) [inline tests/regression/roles/sonic_lag_interfaces/defaults/main.yml:94]: Please change to 'Override'.

File: tests/regression/roles/sonic_lag_interfaces/defaults/main.yml [hunk 3]
```diff
@@ -77,13 +79,49 @@ tests:
...
+        - name: portchannel 40
+          members:
+            interfaces:
+              - member: "{{ interface1 }}"
+        - name: po41
+          members:
+            interfaces:
+              - member: "{{ interface2 }}"
+  - name: test_case_08
+    description: Overridden all portchannel configuration
```
Comment (ArunSaravananBalachandran) [inline tests/regression/roles/sonic_lag_interfaces/defaults/main.yml:106]: Please change to 'Override'.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py [hunk 4]
```diff
@@ -208,20 +208,30 @@ def _state_overridden(self, want, have, diff_members, diff_portchannels):
...
+        requests_deleted_po = self.get_delete_portchannel_requests(deleted_po_list)
+        requests.extend(requests_deleted_po)
+        commands.extend(deleted_po_list)
```
Comment (ArunSaravananBalachandran) [inline plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:229]: Please update the deleted members and port_channel commands with 'deleted' state.

**Review by ArunSaravananBalachandran — CHANGES_REQUESTED** [review]

**Review by ArunSaravananBalachandran — CHANGES_REQUESTED** [review]
Please add changelog fragment for this PR.

File: plugins/modules/sonic_lag_interfaces.py [hunk 5]
```diff
@@ -124,6 +126,94 @@
...
+# interface Eth1/6
+#   channel-group 20
+#   mtu 9100
+#   speed 100000
+#   no shutdown
+# interface Eth1/5
+#   no channel-group
+#   mtu 9100
+#   speed 100000
+#   no shutdown
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_lag_interfaces.py]: Done.

File: plugins/modules/sonic_lag_interfaces.py [hunk 6]
```diff
@@ -124,6 +126,94 @@
...
+# interface Eth1/6
+#   channel-group 20
+#   mtu 9100
+#   speed 100000
+#   no shutdown
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_lag_interfaces.py]: Done.

File: plugins/modules/sonic_lag_interfaces.py [hunk 7]
```diff
@@ -124,6 +126,94 @@
...
+#
+# Before state:
+# -------------
+#
+# interface Eth1/5
+#   channel-group 10
+#   mtu 9100
+#   speed 100000
+#   no shutdown
+# interface Eth1/7
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_lag_interfaces.py]: Done.

File: plugins/modules/sonic_lag_interfaces.py [hunk 8]
```diff
@@ -124,6 +126,94 @@
...
+#   speed 100000
+#   no shutdown
+#
+# interface Eth1/6
+#   channel-group 20
+#   mtu 9100
+#   speed 100000
+#   no shutdown
+#
+- name: Replace device configuration of specified LAG attributes
```
[the remark this answers is not in the document]
Comment (mingjunzhang2019) [reply plugins/modules/sonic_lag_interfaces.py]: Done.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 1 above)
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:226]: Thanks. This is in existing code logic. "replaced_list" is from  "delete_members" and "deleted_po_list" is from "want". Let's not change it for now.

File: tests/regression/roles/sonic_lag_interfaces/defaults/main.yml (hunk 2 above)
Comment (mingjunzhang2019) [reply tests/regression/roles/sonic_lag_interfaces/defaults/main.yml]: Changed.

File: tests/regression/roles/sonic_lag_interfaces/defaults/main.yml (hunk 3 above)
Comment (mingjunzhang2019) [reply tests/regression/roles/sonic_lag_interfaces/defaults/main.yml]: Changed.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 4 above)
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: Changed.

**Comment by mingjunzhang2019** [comment]
> Please add changelog fragment for this PR.

changelog fragment for this PR has been added in a seperated PR for general fragments.

File: plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py (hunk 1 above)
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py:226]: In current implementation, both 'replaced_list' and 'deleted_po_list' are derived from 'have':
- have->delete_list->delete_members->replaced_list
- have->delete_list->deleted_po_list

Its okay to have the existing logic in place for now and the simplification to be performed later.

Please update the deleted members commands with 'deleted' state. (line 219)

Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/lag_interfaces/lag_interfaces.py]: Addressed. Thanks!

**Pull request merged**